### PR TITLE
Refactor/trade book

### PR DESCRIPTION
**Pull request by arealclimber**
### DEVELOP

- [x] marketContext
- [x] src/lib/books/trade_book.ts

### CHECK LIST
 
- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 0
- new Class / Component:  0
- new loop: 0
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/Global-Context-public-d93ac7c941c94fb6b173d34a7bc4e959)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/lib/books/trade_book.ts [hunk 1]
```diff
@@ -319,61 +367,71 @@ class TradeBook {
     this.config = config;
   }
 
-  listTrades() {
-    return this.predictedTrades;
+  listTrades(ticker: string) {
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:370]: 要確保 ticker 有對應的 key

File: src/lib/books/trade_book.ts [hunk 2]
```diff
@@ -150,22 +156,52 @@ class TradeBook {
       ).sort((a, b) => +a.tradeId - +b.tradeId);
     }
   }
+  */
+
+  private _trim(ticker: string): void {
+    const trades = this.trades.get(ticker)!;
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:162]: `this.ensureTickerExists()`

File: src/lib/books/trade_book.ts [hunk 3]
```diff
@@ -150,22 +156,52 @@ class TradeBook {
       ).sort((a, b) => +a.tradeId - +b.tradeId);
     }
   }
+  */
+
+  private _trim(ticker: string): void {
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:161]: - 在每次 get 之前都 `this.ensureTickerExists()`
- 或者用 decorator
- 或者用 prototype 改 .get()

File: src/lib/books/trade_book.ts [hunk 4]
```diff
@@ -150,22 +156,52 @@ class TradeBook {
...
+    const trades = this.trades.get(ticker)!;
+    const predictedTrades = this.predictedTrades.get(ticker)!;
+
+    const now = Date.now();
+    const cutoffTime = now - this.config.holdingTradesMs;
+
+    const trimmedTrades = trades
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:168]: ~用 `.findIndex()`~
```
.reduce((acc, cur) => {
            // Info: remove duplicate tradeId (20230601 - Tzuhan)
            acc[cur.tradeId] = cur;
            return acc;
          }, {} as {[key: string]: ITrade})
```

File: src/lib/books/trade_book.ts [hunk 5]
```diff
@@ -150,22 +156,52 @@ class TradeBook {
...
+      .filter(trade => trade.timestampMs >= cutoffTime)
```
Comment (arealclimber) [inline src/lib/books/trade_book.ts:169]: ~用 `.findIndex()`~
```
.reduce((acc, cur) => {
            // Info: remove duplicate tradeId (20230601 - Tzuhan)
            acc[cur.tradeId] = cur;
            return acc;
          }, {} as {[key: string]: ITrade})
```

File: src/contexts/market_context.tsx [hunk 6]
```diff
@@ -595,8 +600,9 @@ export const MarketProvider = ({children}: IMarketProvider) => {
           timestampMs: trade.timestamp,
           quantity: trade.amount,
         }));
+        const target = trades[0].targetAsset;
```
Comment (arealclimber) [inline src/contexts/market_context.tsx:603]: 要用 `instId`

File: src/contexts/market_context.tsx [hunk 7]
```diff
@@ -745,7 +755,7 @@ export const MarketProvider = ({children}: IMarketProvider) => {
     () =>
       notificationCtx.emitter.on(TideBitEvent.TRADES, (action: IPusherAction, trade: ITrade) => {
         if (trade.instId === selectedTickerRef.current?.instId) {
-          tradeBook.add({
+          tradeBook.add(trade.baseUnit, {
```
Comment (arealclimber) [inline src/contexts/market_context.tsx:758]: 要用 `instId`

File: src/contexts/market_context.tsx [hunk 8]
```diff
@@ -610,7 +616,7 @@ export const MarketProvider = ({children}: IMarketProvider) => {
     return result;
   };
 
-  const syncCandlestickData = (timeSpan?: ITimeSpanUnion) => {
+  const syncCandlestickData = (tickerId: ICurrency, timeSpan?: ITimeSpanUnion) => {
```
Comment (arealclimber) [inline src/contexts/market_context.tsx:619]: 要用 `instId`

File: src/contexts/market_context.tsx [hunk 9]
```diff
@@ -621,7 +627,11 @@ export const MarketProvider = ({children}: IMarketProvider) => {
       if (timeSpan) setTimeSpan(timeSpan);
       const interval = Math.round(getTime(ts) / CANDLESTICK_SIZE);
 
-      const candlesticks = tradeBook.toCandlestick(millesecondsToSeconds(getTime(ts)), 100);
+      const candlesticks = tradeBook.toCandlestick(
+        tickerId,
```
Comment (arealclimber) [inline src/contexts/market_context.tsx:631]: 要用 `instId`

**Review by TzuHanLiang — COMMENTED** [review]
- looked

**Review by Luphia — APPROVED** [review]
- no bug found

**Review by Julian0701 — APPROVED** [review]

**Pull request merged**